File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int product = 1;
        int n = nums.size();
        vector<int> zero_idx;
        vector<int> answer(nums.size(), 0);
        for(int i=0; i<n; i++)
        {
            if(nums[i] != 0) product *= nums[i];
            else zero_idx.push_back(i);
        }
        if(zero_idx.size() == 0)
        {
            for(int i=0; i<n; i++) answer[i] = product / nums[i]; 
        }
        else if(zero_idx.size() == 1)
        {
            answer[zero_idx[0]] = product;
        }
        return answer;
    }
};

File: Data Structures & Algorithms/first-missing-positive/submission-4.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        i = 0
        while i<len(nums):
            if 0<nums[i]<=len(nums) and nums[i] != i+1 and nums[nums[i]-1] != nums[i]:
                nums[nums[i]-1], nums[i] = nums[i], nums[nums[i]-1]
            else:
                i += 1
        for i, n in enumerate(nums):
            if i+1 != n:
                return i+1
        return nums[-1]+1
